sql/catalyst/src/main/java/org/apache/spark/sql/SqlIdentifierUtil.java

```diff
@@ -0,0 +1,322 @@
+ * from org.apache.derby.iapi.util.IdUtil and
+ * org.apache.derby.iapi.util.StringUtil.
+ */
+public class SqlIdentifierUtil {
+
+  /**
+   * Quote a string so that it can be used as an identifier or a string
+   * literal in SQL statements. Identifiers are surrounded by double quotes
+   * and string literals are surrounded by single quotes. If the string
+   * contains quote characters, they are escaped.
+   *
+   * @param source the string to quote
+   * @param quote  the character to quote the string with (' or &quot;)
+   * @return a string quoted with the specified quote character
+   */
+  public static String quoteString(String source, char quote) {
+    // Normally, the quoted string is two characters longer than the source
+    // string (because of start quote and end quote).
+    StringBuffer quoted = new StringBuffer(source.length() + 2);
+    quoted.append(quote);
+    for (int i = 0; i < source.length(); i++) {
+      char c = source.charAt(i);
+      // if the character is a quote, escape it with an extra quote
+      if (c == quote) quoted.append(quote);
+      quoted.append(c);
+    }
+    quoted.append(quote);
+    return quoted.toString();
+  }
+
+  /**
+   * Parse a multi-part (dot separated) SQL identifier from the
+   * String provided. Raise an excepion
+   * if the string does not contain valid SQL indentifiers.
+   * The returned String array contains the normalized form of the
+   * identifiers.
+   *
+   * @param s        The string to be parsed
+   * @param quoteCharacter The character which frames a delimited id (e.g., " or `)
+   * @param upperCaseIdentifiers True if SQL ids are normalized to upper case.
+   * @return An array of strings made by breaking the input string at its dots, '.'.
+   * @throws SparkException Invalid SQL identifier.
+   */
+  public static String[] parseMultiPartSQLIdentifier(
+      String s,
+      char quoteCharacter,
+      boolean upperCaseIdentifiers)
+      throws SparkException {
+    StringReader r = new StringReader(s);
+    String[] qName = parseMultiPartSQLIdentifier(
+        s,
+        r,
+        quoteCharacter,
+        upperCaseIdentifiers);
+    verifyEmpty(s, r);
+    return qName;
+  }
+
+  /**
+   * Parse a multi-part (dot separated) SQL identifier from the
+   * String provided. Raise an excepion
+   * if the string does not contain valid SQL indentifiers.
+   * The returned String array contains the normalized form of the
+   * identifiers.
+   *
+   * @param orig The full text being parsed
+   * @param r        The multi-part identifier to be parsed
+   * @param quoteCharacter The character which frames a delimited id (e.g., " or `)
+   * @param upperCaseIdentifiers True if SQL ids are normalized to upper case.
+   * @return An array of strings made by breaking the input string at its dots, '.'.
+   * @throws SparkException Invalid SQL identifier.
+   */
+  private static String[] parseMultiPartSQLIdentifier(
```

Comment:
A lot of this looks accomplishable more simply with regexes. Is that not feasible?